docs/dcp2_system_design.rst

```diff
@@ -986,6 +1065,400 @@ Broad Institute).
+imported. The *original* is the copy that was most recently imported into TDR,
+the one that's currently present in the TDR dataset. Under this definition, a
+deletion or removal marker object is considerd an altered copy. It is
+insignificant if the original was imported from a staging area at the same URI
+as the one containing the altered copy, or a different URI.
+
+We further differentiate between removals and deletions. Data files can be
+deleted or removed, subgraphs and metadata entities can only be removed. A
+removed data file, subgraph or metadata entity disappears from subsequently
+created snapshots but remains present in previously created snapshots. A
+deleted data file is both removed from subsequently created snapshots **and**
+erased physically from storage, references to it in previously created
+snapshots become invalid and requesting the data file's content results in an
+error.
+
+
+Update an entity
+~~~~~~~~~~~~~~~~
+
+Write the updated entity document to an object in the ``metadata`` directory
+of the staging area. The name of that object must have the same
+``entity_type`` and ``entity_id`` as the original and a strictly higher
+``version``.
+
+
+Rename or update a data file
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+1. Write the updated data file to an object in the ``data`` directory of the
+   staging area, using the same or a different file name as the original data
+   file.
+
+2. Write the updated descriptor to an object in the ``descriptors`` directory
+   of the staging area. The name of the descriptor object must have the same
+   ``entity_type`` and ``entity_id`` as the original, and a strictly higher
+   ``version``. The ``file_id`` property of the descriptor document must be
+   the same as that of the original, and the ``file_version`` property must be
+   strictly higher. The ``file_name`` property can have the same value or a
+   different one. Note that the ``file_id``property of the descriptor document
+   is independent from the column of the same name in TDR ``…_file`` tables and
+   that this section only applies to the former.
+
+   If the ``file_name`` property is the same as in the original and the
+   staging area has the `is_delta`` property set to ``true`` (``false``), the
+   ``sha1`` and ``sha256`` properties must (should) differ from their
+   respective values in the orginal. [#]_ In other words, redundant updates
+   are disallowed in delta staging areas, but discouraged otherwise.
+
+   Similarly, if the ``size`` property is different, the ``sha1`` and
+   ``sha256`` properties must differ between original and update. [#]_ The
+   ``content-type`` property may have the same value or a different one. For
+   updating *only* the content type of a file see `Update the content type of
+   a data file`_.
+
+3. Update the entity describing the data file, as specified in Update an
+   entity. Do so even if the entity document itself did not change. This is
```

Comment:
> Do so even if the entity document itself did not change. 

Meaning the entity should be included with a new name (as described below) even if the contents of the object remain the same?

Reply:
That's what I thought. Maybe we should disallow the unchanged case for delta staging areas, since it now technically contradicts the constraint that delta SAs not contain unchanged entities.

@aherbst-broad

Reply:
The existence of the descriptor is a hack that we did to avoid having to update the file core schema during MVP. I regret agreeing with it but it is what it is. It helps to view descriptor and entity as one inseparable pair.

> But do we still need to include the actual data file if it hasn't changed?

If you are referring to the renaming case then yes, you would have to supply the data file with the same content but under a new name. I do realize that that's an unnecessary burden. Again, I would defer to @aherbst-broad here. If he is amenable to an optimization of eliding the data file if only its name changes then I will write this in. We could also write it in and mark it for later implementation, so that the initial implementation would not allow renaming data files. Maybe that deferment would sweeten the deal for Andrew.

For other alterations, like just changing the metadata entity that describes a data file, you would not need to supply the unchanged data file.

Reply:
We can't also handle file renaming in Ingest yet. 

> For other alterations, like just changing the metadata entity that describes a data file, you would not need to supply the unchanged data file.

Cool, this works for us!

Reply:
The current join point we use to inject file_ids into the file metadata tables is file_id + file_name + crc32. If the file name changes we need to reingest so the join is successful, otherwise we'll end up with a null file ID.

I am ok deferring, but would like to exclude that for now.

Reply:
I removed renaming of data files from the spec.

Reply:
Addressed.